In [10]:
import pandas as pd

In [34]:
# Read data
df = pd.read_pickle('./data/df.pkl')

In [35]:
df.head()

country         state section  \
0  Argentina  Buenos Aires   Total   
1  Argentina  Buenos Aires   Total   
2  Argentina  Buenos Aires   Total   
3  Argentina  Buenos Aires   Total   
4  Argentina  Buenos Aires   Total   

                                            party    votes  percentage  \
0                            Juntos por el cambio  3480298      0.3981   
1                                 Frente de todos  3368310      0.3853   
2                                 Avanza libertad   656498      0.0750   
3  Frente de izquierda y de trabajadores - unidad   596723      0.0682   
4                            Frente vamos con vos   382445      0.0437   

                                   party_original       governor party_spec  \
0                                          Juntos  Axel Kicillof         PJ   
1                                 Frente de todos  Axel Kicillof         PJ   
2                                 Avanza libertad  Axel Kicillof         PJ   
3  Frente de izquierda y de trabajadores - unidad  Axel Kicillof         PJ   
4                            Frente vamos con vos  Axel Kicillof         PJ   

   party_governors party_original_governors  
0  Frente de todos          Frente de todos  
1  Frente de todos          Frente de todos  
2  Frente de todos          Frente de todos  
3  Frente de todos          Frente de todos  
4  Frente de todos          Frente de todos

In [54]:
# Create states dataframe
states_mask = df['section'] == 'Total'
df_states = df[states_mask]
df_states['party_rank_desc'] = df_states.groupby(['party'])['percentage'].rank(ascending=False)
df_states['party_rank_asc'] = df_states.groupby(['party'])['percentage'].rank()
df_states['state_rank_desc'] = df_states.groupby(['state'])['percentage'].rank(ascending=False)
df_states['state_rank_asc'] = df_states.groupby(['state'])['percentage'].rank()


<ipython-input-54-00564a0e1eaf>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_states['party_rank_desc'] = df_states.groupby(['party'])['percentage'].rank(ascending=False)
<ipython-input-54-00564a0e1eaf>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_states['party_rank_asc'] = df_states.groupby(['party'])['percentage'].rank()
<ipython-input-54-00564a0e1eaf>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value

In [69]:
# Votes per party
df_states.groupby('party')['votes'].sum().reset_index().sort_values(by='votes', ascending=False).head(5)

party    votes
29                            Juntos por el cambio  9408079
18                                 Frente de todos  7678354
17  Frente de izquierda y de trabajadores - unidad  1194947
2                                  Avanza libertad  1011076
27                             Hacemos por córdoba   490235

In [63]:
# Top performers
df_states.sort_values(by='percentage', ascending=False).head(3)

country                state section                 party   votes  \
2886  Argentina  Santiago del Estero   Total       Frente de todos  338599   
1233  Argentina           Corrientes   Total  Juntos por el cambio  318542   
1652  Argentina              Formosa   Total       Frente de todos  175171   

      percentage              party_original        governor  \
2886      0.6350  Frente civico por santiago  Gerardo Zamora   
1233      0.5891      Eco + vamos corrientes  Gustavo Valdés   
1652      0.5764             Frente de todos   Gildo Insfrán   

                      party_spec       party_governors  \
2886  Frente Cívico por Santiago       Frente de todos   
1233                         UCR  Juntos por el cambio   
1652                          PJ       Frente de todos   

      party_original_governors  party_rank_desc  party_rank_asc  \
2886           Frente de todos              1.0            23.0   
1233  Encuentro por corrientes              1.0            20.0   
1652     Frente de la victoria              2.0            22.0   

      state_rank_desc  state_rank_asc  
2886              1.0             7.0  
1233              1.0             4.0  
1652              1.0             3.0

In [74]:
# Worse performance of important parties
important_parties = ['Juntos por el cambio', 'Frente de todos', 'Frente de izquierda y de trabajadores - unidad', 'Avanza libertad']
important_parties = ['Juntos por el cambio', 'Frente de todos']
important_parties_mask = df_states['party'].isin(important_parties)
df_states[important_parties_mask].sort_values(by='percentage').head(3)


country                state section                 party   votes  \
1339  Argentina              Córdoba   Total       Frente de todos  205581   
2887  Argentina  Santiago del Estero   Total  Juntos por el cambio   71837   
2101  Argentina             Misiones   Total       Frente de todos   95441   

      percentage               party_original             governor  \
1339      0.1048              Frente de todos      Juan Schiaretti   
2887      0.1347  Frente juntos por el cambio       Gerardo Zamora   
2101      0.1524              Frente de todos  Oscar Herrera Ahuad   

                      party_spec                   party_governors  \
1339                          PJ               Hacemos por córdoba   
2887  Frente Cívico por Santiago                   Frente de todos   
2101                         PCS  Frente renovador de la concordia   

              party_original_governors  party_rank_desc  party_rank_asc  \
1339               Hacemos por córdoba             23.0             1.0   
2887                   Frente de todos             20.0             1.0   
2101  Frente renovador de la concordia             22.0             2.0   

      state_rank_desc  state_rank_asc  
1339              3.0             5.0  
2887              2.0             6.0  
2101              3.0             3.0

In [65]:
# Top states per party
party_rank_mask = df_states['party_rank_desc'] < 6
important_parties_mask = df_states['party'].isin(important_parties)
df_states[important_parties_mask & party_rank_mask].sort_values(by=['party', 'party_rank_desc'])

country                state section                 party    votes  \
2886  Argentina  Santiago del Estero   Total       Frente de todos   338599   
1652  Argentina              Formosa   Total       Frente de todos   175171   
1848  Argentina             La Rioja   Total       Frente de todos    98303   
858   Argentina            Catamarca   Total       Frente de todos   101515   
943   Argentina                Chaco   Total       Frente de todos   263929   
1233  Argentina           Corrientes   Total  Juntos por el cambio   318542   
1526  Argentina           Entre Ríos   Total  Juntos por el cambio   432314   
1337  Argentina              Córdoba   Total  Juntos por el cambio  1059399   
1938  Argentina              Mendoza   Total  Juntos por el cambio   486149   
1682  Argentina                Jujuy   Total  Juntos por el cambio   196293   

      percentage              party_original          governor  \
2886      0.6350  Frente civico por santiago    Gerardo Zamora   
1652      0.5764             Frente de todos     Gildo Insfrán   
1848      0.5606             Frente de todos  Ricardo Quintela   
858       0.5056             Frente de todos        Raúl Jalil   
943       0.4433             Frente de todos  Jorge Capitanich   
1233      0.5891      Eco + vamos corrientes    Gustavo Valdés   
1526      0.5452       Juntos por entre ríos    Gustavo Bordet   
1337      0.5404        Juntos por el cambio   Juan Schiaretti   
1938      0.4954              Cambia mendoza    Rodolfo Suárez   
1682      0.4905                Cambia jujuy   Gerardo Morales   

                      party_spec       party_governors  \
2886  Frente Cívico por Santiago       Frente de todos   
1652                          PJ       Frente de todos   
1848                          PJ       Frente de todos   
858                           PJ       Frente de todos   
943                           PJ       Frente chaqueño   
1233                         UCR  Juntos por el cambio   
1526                          PJ       Frente de todos   
1337                          PJ   Hacemos por córdoba   
1938                         UCR  Juntos por el cambio   
1682                         UCR  Juntos por el cambio   

      party_original_governors  party_rank_desc  party_rank_asc  \
2886           Frente de todos              1.0            23.0   
1652     Frente de la victoria              2.0            22.0   
1848           Frente de todos              3.0            21.0   
858            Frente de todos              4.0            20.0   
943            Frente chaqueño              5.0            19.0   
1233  Encuentro por corrientes              1.0            20.0   
1526          Creer entre ríos              2.0            19.0   
1337       Hacemos por córdoba              3.0            18.0   
1938            Cambia mendoza              4.0            17.0   
1682              Cambia jujuy              5.0            16.0   

      state_rank_desc  state_rank_asc  
2886              1.0             7.0  
1652              1.0             3.0  
1848              1.0             5.0  
858               1.0             5.0  
943               1.0             5.0  
1233              1.0             4.0  
1526              1.0             7.0  
1337              1.0             7.0  
1938              1.0             7.0  
1682              1.0             3.0

In [61]:
# Performance officialismo
mask_party = df_states['party'] == df_states['party_governors']
df_officialismo = df_states[mask_party]
df_officialismo.sort_values(by=['percentage'])

country                state section  \
1338  Argentina              Córdoba   Total   
2668  Argentina           Santa Cruz   Total   
2707  Argentina             Santa Fe   Total   
1527  Argentina           Entre Ríos   Total   
2100  Argentina             Misiones   Total   
1     Argentina         Buenos Aires   Total   
3082  Argentina     Tierra del Fuego   Total   
3117  Argentina              Tucumán   Total   
1734  Argentina             La Pampa   Total   
2540  Argentina             San Juan   Total   
1153  Argentina                 CABA   Total   
1682  Argentina                Jujuy   Total   
1938  Argentina              Mendoza   Total   
858   Argentina            Catamarca   Total   
1848  Argentina             La Rioja   Total   
1652  Argentina              Formosa   Total   
1233  Argentina           Corrientes   Total   
2886  Argentina  Santiago del Estero   Total   

                                 party    votes  percentage  \
1338               Hacemos por córdoba   490235      0.2500   
2668                   Frente de todos    45292      0.2755   
2707                   Frente de todos   563323      0.3130   
1527                   Frente de todos   250895      0.3164   
2100  Frente renovador de la concordia   229225      0.3662   
1                      Frente de todos  3368310      0.3853   
3082                   Frente de todos    37662      0.3967   
3117                   Frente de todos   406411      0.4215   
1734                   Frente de todos    89409      0.4239   
2540                   Frente de todos   176969      0.4346   
1153              Juntos por el cambio   855562      0.4701   
1682              Juntos por el cambio   196293      0.4905   
1938              Juntos por el cambio   486149      0.4954   
858                    Frente de todos   101515      0.5056   
1848                   Frente de todos    98303      0.5606   
1652                   Frente de todos   175171      0.5764   
1233              Juntos por el cambio   318542      0.5891   
2886                   Frente de todos   338599      0.6350   

                        party_original                         governor  \
1338               Hacemos por córdoba                  Juan Schiaretti   
2668                   Frente de todos                  Alicia Kirchner   
2707                   Frente de todos                     Omar Perotti   
1527                   Frente de todos                   Gustavo Bordet   
2100  Frente renovador de la concordia              Oscar Herrera Ahuad   
1                      Frente de todos                    Axel Kicillof   
3082                   Frente de todos                  Gustavo Melella   
3117                   Frente de todos  Juan Manzur(En uso de licencia)   
1734                   Frente de todos                  Sergio Ziliotto   
2540  Frente de todos - todos san juan                      Sergio Uñac   
1153              Juntos por el cambio        Horacio Rodríguez Larreta   
1682                      Cambia jujuy                  Gerardo Morales   
1938                    Cambia mendoza                   Rodolfo Suárez   
858                    Frente de todos                       Raúl Jalil   
1848                   Frente de todos                 Ricardo Quintela   
1652                   Frente de todos                    Gildo Insfrán   
1233            Eco + vamos corrientes                   Gustavo Valdés   
2886        Frente civico por santiago                   Gerardo Zamora   

                      party_spec                   party_governors  \
1338                          PJ               Hacemos por córdoba   
2668                      Kolina                   Frente de todos   
2707                          PJ                   Frente de todos   
1527                          PJ                   Frente de todos   
2100                         PCS  Frente renovador de la concordia   
1                             PJ                   

In [53]:
# Best performance no oficialismo
mask_party = df_states['party'] != df_states['party_governors']
df_no_officialismo = df_states[mask_party]
df_no_officialismo['party_rank_desc'] = df_no_officialismo.groupby(['party'])['percentage'].rank(ascending=False)
party_rank_mask = df_no_officialismo['party_rank_desc'] < 4
df_no_officialismo[party_rank_mask].sort_values(by=['party', 'party_rank_desc'])

<ipython-input-53-061e7bd39708>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_no_officialismo['party_rank_desc'] = df_no_officialismo.groupby(['party'])['percentage'].rank(ascending=False)


country         state section                         party   votes  \
5     Argentina  Buenos Aires   Total                     + valores  257568   
1157  Argentina          CABA   Total  Autodeterminación y libertad   55965   
1155  Argentina          CABA   Total               Avanza libertad  310036   
2     Argentina  Buenos Aires   Total               Avanza libertad  656498   
1342  Argentina       Córdoba   Total               Avanza libertad   44542   
...         ...           ...     ...                           ...     ...   
1343  Argentina       Córdoba   Total         Unión popular federal   16341   
1942  Argentina       Mendoza   Total            Vamos ! mendocinos   38038   
1235  Argentina    Corrientes   Total                 Vamos con vos   13766   
1849  Argentina      La Rioja   Total                Vamos la rioja   49077   
1940  Argentina       Mendoza   Total                         Verde   90671   

      percentage                party_original                   governor  \
5         0.0294                     + valores              Axel Kicillof   
1157      0.0307  Autodeterminación y libertad  Horacio Rodríguez Larreta   
1155      0.1703            La libertad avanza  Horacio Rodríguez Larreta   
2         0.0750               Avanza libertad              Axel Kicillof   
1342      0.0227            La libertad avanza            Juan Schiaretti   
...          ...                           ...                        ...   
1343      0.0083         Unión popular federal            Juan Schiaretti   
1942      0.0387            Vamos ! mendocinos             Rodolfo Suárez   
1235      0.0254                 Vamos con vos             Gustavo Valdés   
1849      0.2799                Vamos la rioja           Ricardo Quintela   
1940      0.0923                         Verde             Rodolfo Suárez   

     party_spec       party_governors  party_original_governors  \
5            PJ       Frente de todos           Frente de todos   
1157        PRO  Juntos por el cambio      Juntos por el cambio   
1155        PRO  Juntos por el cambio      Juntos por el cambio   
2            PJ       Frente de todos           Frente de todos   
1342         PJ   Hacemos por córdoba       Hacemos por córdoba   
...         ...                   ...                       ...   
1343         PJ   Hacemos por córdoba       Hacemos por córdoba   
1942        UCR  Juntos por el cambio            Cambia mendoza   
1235        UCR  Juntos por el cambio  Encuentro por corrientes   
1849         PJ       Frente de todos           Frente de todos   
1940        UCR  Juntos por el cambio            Cambia mendoza   

      party_rank_desc  rank_by_state  
5                 1.0            6.0  
1157              1.0            5.0  
1155              1.0            3.0  
2                 2.0            3.0  
1342              3.0            6.0  
...               ...            ...  
1343              1.0            7.0  
1942              1.0            5.0  
1235              1.0            3.0  
1849              1.0            2.0  
1940              1.0            3.0  

[78 rows x 13 columns]

In [52]:
# Diferencia entre los dos primeros
df_states['rank_by_state'] = df_states.groupby('state')['percentage'].rank(ascending=False)
df_states_first_two = df_states[df_states['rank_by_state'] <= 2]
df_states_first_two.sort_values(by=['state', 'percentage'], inplace=True)
df_states_first_two['perc_diff'] = df_states_first_two['percentage'] - df_states_first_two['percentage'].shift()
df_states_first_two[df_states_first_two['rank_by_state'] == 1].sort_values(by='perc_diff')

<ipython-input-52-118d117a0271>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_states['rank_by_state'] = df_states.groupby('state')['percentage'].rank(ascending=False)
<ipython-input-52-118d117a0271>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_states_first_two.sort_values(by=['state', 'percentage'], inplace=True)
<ipython-input-52-118d117a0271>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pyda

country                state section                        party  \
2616  Argentina             San Luis   Total          Unidos por san luis   
2540  Argentina             San Juan   Total              Frente de todos   
0     Argentina         Buenos Aires   Total         Juntos por el cambio   
943   Argentina                Chaco   Total              Frente de todos   
3117  Argentina              Tucumán   Total              Frente de todos   
2372  Argentina                Salta   Total              Frente de todos   
2099  Argentina             Misiones   Total         Juntos por el cambio   
1733  Argentina             La Pampa   Total         Juntos por el cambio   
2189  Argentina              Neuquén   Total  Movimiento popular neuquino   
2666  Argentina           Santa Cruz   Total         Juntos por el cambio   
2706  Argentina             Santa Fe   Total         Juntos por el cambio   
1073  Argentina               Chubut   Total         Juntos por el cambio   
2294  Argentina            Río Negro   Total       Juntos somos rio negro   
3082  Argentina     Tierra del Fuego   Total              Frente de todos   
858   Argentina            Catamarca   Total              Frente de todos   
1652  Argentina              Formosa   Total              Frente de todos   
1153  Argentina                 CABA   Total         Juntos por el cambio   
1233  Argentina           Corrientes   Total         Juntos por el cambio   
1526  Argentina           Entre Ríos   Total         Juntos por el cambio   
1682  Argentina                Jujuy   Total         Juntos por el cambio   
1938  Argentina              Mendoza   Total         Juntos por el cambio   
1848  Argentina             La Rioja   Total              Frente de todos   
1337  Argentina              Córdoba   Total         Juntos por el cambio   
2886  Argentina  Santiago del Estero   Total              Frente de todos   

        votes  percentage                    party_original  \
2616   124837      0.4607               Unidos por san luis   
2540   176969      0.4346  Frente de todos - todos san juan   
0     3480298      0.3981                            Juntos   
943    263929      0.4433                   Frente de todos   
3117   406411      0.4215                   Frente de todos   
2372   200730      0.3237                   Frente de todos   
2099   256090      0.4091       Frente juntos por el cambio   
1733   101259      0.4801              Juntos por el cambio   
2189   111177      0.2942       Movimiento popular neuquino   
2666    57571      0.3502                 Cambia santa cruz   
2706   724008      0.4023              Juntos por el cambio   
1073   108556      0.3795       Juntos por el cambio chubut   
2294   138545      0.3727            Juntos somos rio negro   
3082    37662      0.3967                   Frente de todos   
858    101515      0.5056                   Frente de todos   
1652   175171      0.5764                   Frente de todos   
1153   855562      0.4701              Juntos por el cambio   
1233   318542      0.5891            Eco + vamos corrientes   
1526   432314      0.5452             Juntos por entre ríos   
1682   196293      0.4905                      Cambia jujuy   
1938   486149      0.4954                    Cambia mendoza   
1848    98303      0.5606                   Frente de todos   
1337  1059399      0.5404              Juntos por el cambio   
2886   338599      0.6350        Frente civico por santiago   

                             governor                  party_spec  \
2616            Alberto Rodríguez Saá                          PJ   
2540                      Sergio Uñac                          PJ   
0                       Axel Kicillof                          PJ   
943                  Jorge Capitanich                          PJ   
3117  Juan Manzur(En uso de licencia)                          PJ   
2372                    Gustavo Sáenz                        PAIS   
2099              Oscar

In [72]:
# Foco Avanza libertad
df_states[df_states['party'] == 'Avanza libertad']

country         state section            party   votes  percentage  \
2     Argentina  Buenos Aires   Total  Avanza libertad  656498      0.0750   
1155  Argentina          CABA   Total  Avanza libertad  310036      0.1703   
1342  Argentina       Córdoba   Total  Avanza libertad   44542      0.0227   

          party_original                   governor party_spec  \
2        Avanza libertad              Axel Kicillof         PJ   
1155  La libertad avanza  Horacio Rodríguez Larreta        PRO   
1342  La libertad avanza            Juan Schiaretti         PJ   

           party_governors party_original_governors  party_rank_desc  \
2          Frente de todos          Frente de todos              2.0   
1155  Juntos por el cambio     Juntos por el cambio              1.0   
1342   Hacemos por córdoba      Hacemos por córdoba              3.0   

      party_rank_asc  state_rank_desc  state_rank_asc  
2                2.0              3.0             4.0  
1155             3.0              3.0             3.0  
1342             1.0              6.0             2.0



El mas pijudo: Zamora.
El otro: Gustavo Valdes.
La republica amarilla: Cordoba.
Los maz zurdos: Jujuy.
El papelon: Alicia Kirchner en Santa Cruz.
Avanza Libertad: Un fenomeno de la Capital Federal.

Las 3 provincias más K: Santiago del Estero (63,5%), Formosa (57,6%) y La Rioja (56%). 

Acá sus gobernadores Gerardo Zamora, Gildo Insfrán y Ricardo Quintela todos ganadores y sonrientes.

Las 3 provincias donde Juntos por el Cambio tuvo sus mejores resultados: Corrientes (58,9%), Entre Rios (54,5%) y Cordoba (54%).

Abajo el gobernador de Corrientes Gustavo Valdes sonriente ganador, Gustavo Bordet y Juan Schiaretti preocupados por las derrotas en sus provincias.

El incumbente mas malo Alicia Kirchner, con el 27,5% de los votos.

EL incumbente mas larga: Santiago del Estero.
El Juntos por el Cambio mas larga: Cordoba.